token-erc-721/chaincode-javascript/lib/tokenERC721.js

```diff
@@ -0,0 +1,403 @@
+        const iterator = await ctx.stub.getStateByPartialCompositeKey(balancePrefix, [owner]);
+
+        // Count the number of returned composite keys
+        let balance = 0;
+        let result = await iterator.next();
+        while (!result.done) {
+            balance++;
+            result = await iterator.next();
+        }
+        return balance;
+    }
+
+    /**
+     * OwnerOf finds the owner of a non-fungible token
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} tokenId The identifier for a non-fungible token
+     * @returns {String} Return the owner of the non-fungible token
+     */
+    async OwnerOf(ctx, tokenId) {
+        const nft = await this._readNFT(ctx, tokenId);
+        const owner = nft.owner;
+        if (!owner) {
+            throw new Error('No owner is assigned to this token');
+        }
+
+        return owner;
+    }
+
+    /**
+     * TransferFrom transfers the ownership of a non-fungible token
+     * from one owner to another owner
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} from The current owner of the non-fungible token
+     * @param {String} to The new owner
+     * @param {String} tokenId the non-fungible token to transfer
+     * @returns {Boolean} Return whether the transfer was successful or not
+     */
+    async TransferFrom(ctx, from, to, tokenId) {
+        const sender = ctx.clientIdentity.getID();
+
+        const nft = await this._readNFT(ctx, tokenId);
+
+        // Check if the sender is the current owner, an authorized operator,
+        // or the approved client for this non-fungible token.
+        const owner = nft.owner;
+        const tokenApproval = nft.approved;
+        const operatorApproval = await this.IsApprovedForAll(ctx, owner, sender);
+        if (owner !== sender && tokenApproval !== sender && !operatorApproval) {
+            throw new Error('The sender is not allowed to transfer the non-fungible token');
+        }
+
+        // Check if `from` is the current owner
+        if (owner !== from) {
+            throw new Error('The from is not the current owner.');
+        }
+
+        // Clear the approved client for this non-fungible token
+        nft.approved = '';
+
+        // Remove a composite key from the balance of the current owner
+        const balanceKeyFrom = ctx.stub.createCompositeKey(balancePrefix, [from, tokenId]);
+        await ctx.stub.deleteState(balanceKeyFrom);
+
+        // Save a composite key to count the balance of a new owner
+        const balanceKeyTo = ctx.stub.createCompositeKey(balancePrefix, [to, tokenId]);
+        await ctx.stub.putState(balanceKeyTo, Buffer.from('\u0000'));
+
+        // Assign a non-fungible token to the new owner.
+        nft.owner = to;
+        const nftKey = ctx.stub.createCompositeKey(nftPrefix, [tokenId]);
+        await ctx.stub.putState(nftKey, Buffer.from(JSON.stringify(nft)));
```

Comment:
I would do this code block prior to the balance key code blocks. At least for me, it is more logical to first write the main token record, then handle the balance records.

In the comment, instead of 'Assign' , I'd recommend the word 'Overwrite' so that people know we are overwriting the existing token record.